lib/Containers/Enumerate.h

```diff
@@ -0,0 +1,130 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2020 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+#ifndef ARANGODB_CONTAINERS_ENUMERATE_H
+#define ARANGODB_CONTAINERS_ENUMERATE_H
+#include <functional>
+#include <iterator>
+
+
+template <typename T, typename C>
+struct enumerate_wrapper;
+
+/*
+ * enumerate function similar to pythons enumerate.
+ * Usage:
+ *  for (auto [idx, e] : enumerate(v)) {
+ *    // idx is incremented for each element
+ *    // e is the reference type of the iterator of the container v.
+ *  }
+ *
+ * Usable for std::vector, std::list, std::map and std::unordered_map. _should_
+ * work for other containers that provide a LegacyIterator as well.
+ */
+template <typename T, typename C = unsigned int>
+enumerate_wrapper<T, C> enumerate(T&& v, C c = C());
+
+
+
+template <typename T, typename C = unsigned int>
+struct enumerate_iterator;
+
+template <typename T, typename C>
+void swap(enumerate_iterator<T, C>& a, enumerate_iterator<T, C>& b);
+
+template <typename T, typename C>
+struct enumerate_iterator {
+  enumerate_iterator() : _iter(), _c(0) {}
```

Comment:
static_assert(std::is_intergral_v<C>);

Reply:
Why? You could as well use your favorite BigInt class here...

Reply:
Really? I think the size that containers can give back is usually `std::size_t` probably you could drop the template and just use that.